### DISCUSSION TITLE
Development and Implementation of Native SDKs for Key Platforms in Directus

### DISCUSSION THREAD

**Discussion by vamsii777** [directus/directus#20963] [Feature Requests]
### Summary

This feature request focuses on the creation of pre-built, native SDKs for Directus, catering to a range of platforms such as Android, iOS, web, and others. These SDKs would offer developers ready-to-use tools and libraries, streamlining the integration of Directus into various applications.

### Basic Example

For instance, an iOS developer would have access to a Directus iOS SDK, designed with Swift libraries, providing a seamless integration experience with the Directus API. This SDK would include Swift-friendly tools and functions, simplifying tasks such as user authentication, data fetching, and synchronization with the Directus backend.

An example of how the Swift SDK could be used is as follows:

```swift
import DirectusSDK

class DirectusService {
    private var directusClient: DirectusClient

    init() {
        directusClient = DirectusClient(apiEndpoint: "https://your-directus-instance.com")
    }

    func authenticateUser(email: String, password: String, completion: @escaping (Result<DirectusSession, Error>) -> Void) {
        directusClient.authenticate(email: email, password: password) { result in
            completion(result)
        }
    }

    func fetchCollectionData(collectionName: String, completion: @escaping (Result<[YourDataType], Error>) -> Void) {
        directusClient.fetchItems(fromCollection: collectionName) { result in
            completion(result)
        }
    }
}
```

In this example, the `DirectusClient` class simplifies the process of interacting with Directus APIs. The `authenticateUser` method manages user authentication, while `fetchCollectionData` allows for easy data retrieval from a specified collection. This SDK approach streamlines development, enabling iOS developers to focus more on their application's unique features and less on backend integrations.

### Motivation

Developing native SDKs for each platform would significantly reduce development time and complexity. Currently, integrating Directus into different platforms requires custom development, which can be inefficient and error-prone. Pre-built SDKs would simplify this process, making Directus more accessible to a broader range of developers.

### Detailed Design

For the detailed design of the Directus SDKs for all platforms, an SDK generator should be employed to ensure standardization and efficiency. This approach would tailor each SDK to the specific norms and standards of its target platform. The key components of these SDKs would include:

1. **Platform-Specific Libraries and Tools:** Utilize the SDK generator to create libraries that are native to each platform (e.g., Swift for iOS, Kotlin for Android, JavaScript for web). This ensures that the SDKs feel intuitive and familiar to developers on each platform.

2. **Modules for Authentication, Data Handling, and API Communication:** The generator should produce modules that handle common tasks such as user authentication, data retrieval, updating, and synchronization with the Directus API. These modules would abstract the complexity of direct API calls, providing a simpler interface for developers.

3. **Comprehensive Documentation and Sample Code:** For each platform-specific SDK, generate detailed documentation and provide sample code. This is crucial for easing the integration process, enabling developers to quickly understand and utilize the SDK in their projects. 

By leveraging an SDK generator, the process of creating and maintaining these SDKs becomes more streamlined and efficient, ensuring consistency across different platforms while also allowing for platform-specific customizations. This approach would significantly enhance the developer experience and broaden the usage of Directus across various application ecosystems.

### Requirements List

Must: Develop SDKs for key platforms like Android, iOS, and web.
Should: Ensure each SDK is intuitive, well-documented, and easy to integrate.
Could: Offer additional tools or plugins for popular development environments.
Won't: Compromise on the performance and reliability of Directus core functionalities.

### Drawbacks

This initiative requires significant resources for initial development and ongoing maintenance. Keeping SDKs up-to-date with Directus updates and new platform standards is also essential.

### Alternatives

One alternative is to enhance community support and encourage open-source development of these SDKs.

### Adoption Strategy

Introduce these SDKs as complementary tools to Directus, ensuring they are non-intrusive and backward compatible. Provide detailed documentation and support to facilitate adoption among developers.

### Unresolved Questions

How will ongoing maintenance and updates for these SDKs be managed?
What strategies should be adopted to cover a broad range of platforms while ensuring quality and consistency?

**Comment by jetsauce** [comment]
Yes, yes, and yes. I'd donate monies for this. Directus Self-Hosted is a very strong offering compared to Supabase Self-Hosted. After spending a week getting Supabase Self-Hosted operational, there were a plethora of console errors still persisting, which makes sense since the self-hosted Supabase app is purposely built for their Paid Cloud users.

By comparison, I was able to get Directus Self-Hosted operational in 1 hour with no headaches and easy to resolve console errors. It was a much more gratifying developer experience since Directus offers a more advanced feature set being a Headless CMS as well as backend manager.

One of the downsides to switching to Directus was the lack of an iOS SDK since I am attempting to quickly create a Swift native app.

Releasing a iOS SDK and then Android SDK would create a very strong Firebase alternative offering. You'd blow Supabase out of the water -- Battleship sunk.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
